From: Magnus Sandin <magnus.sandin@gmail.com>
Date: Fri Feb  3 07:38:41 +0100
Subject: [PATCH] Handle encrypted zfs root partition

If the root partition is encrypted using zfs
grub-probe will fail to get the zfs pool name since
it doen't support encrypted zfs filesystems
and it will thus return an empty pool name.

If the root filesystem is zfs, then skip grub-probe
altogether and use blkid instead to get the label
which is what grub-probe tries to do but fails
if the zfs filesystem is encrypted.

--- a/util/grub.d/10_linux.in
+++ b/util/grub.d/10_linux.in
@@ -74,7 +74,7 @@
 	    GRUB_CMDLINE_LINUX="rootflags=subvol=${rootsubvol} ${GRUB_CMDLINE_LINUX}"
 	fi;;
     xzfs)
-	rpool=`${grub_probe} --device ${GRUB_DEVICE} --target=fs_label 2>/dev/null || true`
+	rpool="`blkid -o value -s LABEL ${GRUB_DEVICE}`"
 	bootfs="`make_system_path_relative_to_its_root / | sed -e "s,@$,,"`"
 	LINUX_ROOT_DEVICE="ZFS=${rpool}${bootfs%/}"
 	;;
--- a/util/grub.d/20_linux_xen.in
+++ b/util/grub.d/20_linux_xen.in
@@ -79,7 +79,7 @@
 	    GRUB_CMDLINE_LINUX="rootflags=subvol=${rootsubvol} ${GRUB_CMDLINE_LINUX}"
 	fi;;
     xzfs)
-	rpool=`${grub_probe} --device ${GRUB_DEVICE} --target=fs_label 2>/dev/null || true`
+	rpool="`blkid -o value -s LABEL ${GRUB_DEVICE}`"
 	bootfs="`make_system_path_relative_to_its_root / | sed -e "s,@$,,"`"
 	LINUX_ROOT_DEVICE="ZFS=${rpool}${bootfs%/}"
 	;;
